Add optimistic mode to template switch

akloeckner opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This adds the optimistic mode from template light also to template switch:
- value_template is optional. If None: assume state change was successful.
- restore last known state after restarting hass

This is useful for switches which do not report a state back, but need a script to be switched. For me this is the case with pilight switches having a bad reception. I need to send the switching command multiple times. Template switch sems to be the only integration to suport such scripts. Hence the PR for this component.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
switch:
- platform: template
  switches:
    wohnzimmer_sitzecke:
      friendly_name: Wohnzimmer Sitzecke
      turn_on:
      - service: pilight.send
        data:
          protocol: elro_800_switch
          systemcode: !secret pilight_systemcode_wohnzimmer
          unitcode: 2
          'on': 1
      turn_off:
      - service: pilight.send
        data:
          protocol: elro_800_switch
          systemcode: !secret pilight_systemcode_wohnzimmer
          unitcode: 2
          'off': 1
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: #31466
- This PR is related to issue: none
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/12020

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

akloeckner committed:
Add optimistic mode to template switch:
- value_template is optional. If None: assume state change was successful.
- restore last known state after restarting hass

akloeckner committed:
fix lint errors

akloeckner committed:
fix black

homeassistant commented:
This pull request has been mentioned on **Home Assistant Community**. There might be relevant details there:

https://community.home-assistant.io/t/generic-template-switch-with-assumed-state/170003/2

akloeckner committed:
isort...

akloeckner commented:
I fixed all CI hints I can understand. With the remaining ones, I'll need your help...

akloeckner committed:
try to increase coverage via testing optimistic mode

akloeckner committed:
add test for restore state

akloeckner commented:
I thought, adding tests might help with the CI. It didn't. But doesn't hurt either. So, here they are.

Please advice on the 
```
ERROR
Tests are leaving files behind. Please update the tests to avoid writing any files:
tests/testing_config/.storage/core.restore_state
```

akloeckner commented:
@balloob, is there anything more I can do to advance this PR?

akloeckner commented:
Obviously, I'd still like to see that merged...

akloeckner commented:
Really? This just ports a feature from `light` to `switch`. So, the change is minimal. It's added value. I spent quite some time on it. I asked for help, mentioned maintainers. But all I get as replies is some bot suggesting to simply trash this work.

How this PR is handled does not raise my motivation to contribute in the future. 

Please at least let me know, why there is no progress. I'll help if I can.

tiagofreire-pt commented:
@akloeckner , I also don't know why this may be stalling and willing to have this on my HA instance. But the code may have some issues to be resolved, as the CI seems to post failure on 2 tests.

Could you force a change so those tests could run again, please?

Thanks.

akloeckner commented:
Sure... Now, it seems we have conflicts...

akloeckner commented:
Bump

frenck commented:
@akloeckner Could you make sure the build passes and the merge conflict is resolved? Thanks 👍

akloeckner commented:
@frenck Thanks for picking this up. I have no idea how to make the build pass. See my comment above: https://github.com/home-assistant/core/pull/31637#issuecomment-586788582

Will it help, if I fix the conflicts and then you can check what is going wrong in the build?

frenck commented:
Start with rebasing the PR to solve the merge conflicts. Next, run the tests to see which fail and find out why and adjusts them accordingly to make them pass.

akloeckner commented:
Alright. I'll start with rebasing. But I will not be able to find out why the tests fail. It does so since the beginning and I copied the specific test from another component. I will need your guidance.

I'll report back here, as soon as I have that PR rebased...

akloeckner commented:
So, here is the rebased PR. Running tox (for the one test file I changed) yields:
```
py36 run-test-pre: PYTHONHASHSEED='1367905925'
py36 run-test: commands[0] | pytest --timeout=9 --durations=10 -n auto --dist=loadfile -qq -o console_output_style=count -p no:sugar tests/components/template/test_switch.py
bringing up nodes...
.................                                                         [17/17]
=========================== slowest 10 test durations ============================
0.34s call     tests/components/template/test_switch.py::TestTemplateSwitch::test_template_state_text
0.25s call     tests/components/template/test_switch.py::TestTemplateSwitch::test_missing_on_does_not_create
0.21s teardown tests/components/template/test_switch.py::test_available_template_with_entities[pyloop]
0.18s teardown tests/components/template/test_switch.py::test_invalid_availability_template_keeps_component_available[pyloop]
0.12s call     tests/components/template/test_switch.py::TestTemplateSwitch::test_on_action
0.12s setup    tests/components/template/test_switch.py::TestTemplateSwitch::test_template_state_text
0.08s call     tests/components/template/test_switch.py::TestTemplateSwitch::test_icon_template
0.08s call     tests/components/template/test_switch.py::TestTemplateSwitch::test_entity_picture_template
0.08s call     tests/components/template/test_switch.py::TestTemplateSwitch::test_off_action
0.07s call     tests/components/template/test_switch.py::TestTemplateSwitch::test_off_action_optimistic
py36 run-test: commands[1] | /srv/homeassistant_dev/script/check_dirty

***** ERROR
Tests are leaving files behind. Please update the tests to avoid writing any files:
tests/testing_config/.storage/core.restore_state

ERROR: InvocationError for command /srv/homeassistant_dev/script/check_dirty (exited with code 1)
```

I am pretty sure, it's caused by the  `test_restore_state` I added. I added that one in order to fix another test, which complained about code coverage. And I copied it from somewhere, e.g. here:
https://github.com/home-assistant/core/blob/2af984917e7cf3ee649ed3a531a872e7aaeba713/tests/components/input_boolean/test_init.py#L123

So, I really don't get, why my test leaves files behind and the other one does not...

tiagofreire-pt commented:
This is odd. The tests are failing with domain resolving issues:

```
stderr:
    fatal: unable to access 'https://github.com/adrienverge/yamllint.git/': Could not resolve host: github.com
```

and

```
stderr:
    fatal: unable to access 'https://github.com/codespell-project/codespell/': Could not resolve host: github.com
```

frenck commented:
GitHub had issues tonight, see: <https://githubstatus.com>.

frenck commented:
/AzurePipelines run

akloeckner commented:
PS: The pipeline now also complains about code ceverage in some device tracker file. Which I did not change. So, I am also out of clues for that second error.

frenck commented:
@akloeckner I think you are reading codecov incorrect. The coverage change on patch level can be found here: <https://codecov.io/gh/home-assistant/core/compare/53a91ece4ecb24624e9913f0a3cba292e7303825...3593b07368b55650aa3ed33dc9f0becc49b2cadd/diff>

However, it isn't really relevant until the tests are passing (as the data is now incomplete).

akloeckner commented:
Ah ok. Then let's leave the codecov for later.

For now, I'll need help with the left-over files...

akloeckner commented:
@frenck, any suggestions?

akloeckner commented:
@frenck, will a `os.remove('tests/testing_config/.storage/core.restore_state')` do the trick?

MartinHjelmare commented on tests/components/template/test_switch.py:
```diff
@@ -563,3 +609,38 @@ async def test_invalid_availability_template_keeps_component_available(hass, cap
     assert hass.states.get("switch.test_template_switch").state != STATE_UNAVAILABLE
     assert ("UndefinedError: 'x' is undefined") in caplog.text
+
+    def test_restore_state(self):
```
Please make this test a standalone async pytest test coroutine function instead. It needs to be that for the storage to be cleaned up properly.

See eg:
https://github.com/home-assistant/core/blob/2af984917e7cf3ee649ed3a531a872e7aaeba713/tests/components/input_boolean/test_init.py#L123-L145

akloeckner replied:
I tried that. It was hard work to make this test pass actually, because the `hass.states.get` would return `None` all the time. Until I added the `wait_till_done`. Now, I'm back with the dirty files left behind...

akloeckner committed:
make the restore cache test something async

balloob commented on homeassistant/components/template/switch.py:
```diff
@@ -141,12 +142,19 @@ def template_switch_state_listener(entity, old_state, new_state):
+        await super().async_added_to_hass()
+        state = await self.async_get_last_state()
+        if state:
+            self._state = state.state in ("true", STATE_ON)
```
State will always be `STATE_ON`, it can never be `true`. 
```suggestion
            self._state = state.state == STATE_ON
```

balloob commented on homeassistant/components/template/switch.py:
```diff
@@ -141,12 +142,19 @@ def template_switch_state_listener(entity, old_state, new_state):
+        if state:
+            self._state = state.state in ("true", STATE_ON)
+
         self.hass.bus.async_listen_once(
```
We don't even need to listen for startup if there is no `self._template`. So you can do 

```python
if self._template is None:
    return

@callback
def template_switch_startup(event):
    …

self.hass.bus.async_listen_once(…)
```

akloeckner replied:
except for the state restoring should be run also (and especially) for optimistic mode. I'll move that in front of the `return` statement

balloob commented on homeassistant/components/template/switch.py:
```diff
@@ -184,13 +192,21 @@ def available(self) -> bool:
         await self._on_script.async_run(context=self._context)
+        if self._template is None:
+            self._state = True
+            self.async_schedule_update_ha_state()
```
Since there is nothing to call in `async_update`, just call `self.async_write_ha_state()`.

balloob commented on tests/components/template/test_switch.py:
```diff
@@ -1,12 +1,41 @@
+
+    async def _storage(items=None, config=None):
+        if items is None:
+            hass_storage["switch"] = {
```
This is weird. Switch doesn't store things in storage ?

balloob commented on tests/components/template/test_switch.py:
```diff
@@ -1,12 +1,41 @@
 
 
+@pytest.fixture
+def storage_setup(hass, hass_storage):
```
This fixture is not used and can be removed.

akloeckner replied:
I also din't really know what it does. I just tried to copy the thing from @MartinHjelmare's example...

akloeckner committed:
address @balloob's review

balloob commented on homeassistant/components/template/switch.py:
```diff
@@ -133,6 +134,17 @@ def __init__(
 
+        # restore state after startup
+        await super().async_added_to_hass()
+        state = await self.async_get_last_state()
+        if state:
+            self._state = state.state == STATE_ON
```
Only execute this if `self._template is None`

balloob commented on homeassistant/components/template/switch.py:
```diff
@@ -141,9 +153,10 @@ def template_switch_state_listener(entity, old_state, new_state):
-            async_track_state_change(
-                self.hass, self._entities, template_switch_state_listener
-            )
+            if self._template is not None:
```
No need to check this, since you already return earlier on line 145.

balloob commented on homeassistant/components/template/switch.py:
```diff
@@ -141,9 +153,10 @@ def template_switch_state_listener(entity, old_state, new_state):
+                    self.hass, self._entities, template_switch_state_listener
+                )
 
             self.async_schedule_update_ha_state(True)
```
This is weird, we shouldn't schedule an update. WE should remove this line and instead in the setup phase it should be `async_add_entiites(entities, True)`

akloeckner replied:
As this would change the original component code and as I don't understand exactly what you suggest here, I prefer not to change this. If you still want me to change something, please explain in more detail. Of course, you can also push a quick change to this branch yourself.

balloob replied:
Oh I actually see now that this is called in the state change handler, it's fine there.

akloeckner committed:
address 2nd review

balloob reviewed (approved):
Looks great! 🐬

balloob commented:
Test failed because tests are leaving files behind. Tests inside classes do not have the right fixtures applied, including the one that writes files. So you will need to move all tests that test omitting the value template from classes to function tests.

akloeckner commented:
> So you will need to move all tests that test omitting the value template from classes to function tests.

Isn't this what @martinhjelmare asked for, but what didn't help? The only test that seems to write files is the one for restoring states. And I moved that outside the class. 

Is there another test that I should move?

akloeckner commented:
I tested the tests a bit, and I'm having the impression that even the original tests now write files. Do I have to move *all* tests to functions? 

However, template light also has optimistic mode. And it's tested in a class. But there's no state restoration for light...

https://github.com/akloeckner/home-assistant/blob/65eb9ca213f617b94b0744d0d91c5efdb3da028d/homeassistant/components/template/light.py#L149

https://github.com/akloeckner/home-assistant/blob/65eb9ca213f617b94b0744d0d91c5efdb3da028d/tests/components/template/test_light.py#L25

balloob commented:
There is a single entry in core.restore_state after I run the tests:

```
❯ cat tests/testing_config/.storage/core.restore_state
{
    "data": [
        {
            "last_seen": "2020-07-04T06:34:34.145896+00:00",
            "state": {
                "attributes": {
                    "friendly_name": "test_template_switch"
                },
                "context": {
                    "id": "c78aefb374c441ac9e7e9ae3a84cdede",
                    "parent_id": null,
                    "user_id": null
                },
                "entity_id": "switch.test_template_switch",
                "last_changed": "2020-07-04T06:34:34.142337+00:00",
                "last_updated": "2020-07-04T06:34:34.142337+00:00",
                "state": "on"
            }
        }
    ],
    "key": "core.restore_state",
    "version": 1
}%              
```

akloeckner committed:
change tests to be async (untested tests yet... absurd, huh?)

akloeckner committed:
async def...

akloeckner committed:
async_set

akloeckner committed:
async_all... it's starting to become annoying...

akloeckner committed:
fix black and other stuff...

akloeckner commented:
Thanks for having a look, @balloob. All tests seem to create that entity_id, however...

I thus went with my first guess: The added state restoration is there not only for the new tests but also for the old ones. And the class doesn't get the corresponing fixtures, as @MartinHjelmare pointed out. (Thanks for that, too!) So, I made **all** the tests be async now. Tests are passing now. There are no files left over anymore. 

Since this seems to be also the "new" way to define tests, I consider it an investment in the future.

akloeckner committed:
oups... should commit tox.ini...

akloeckner commented:
Hmm... codecov is complaining again... 

@balloob, I believe, it's due to the state being `set` now without an `update` (one of your review comments). Is it safe to remove the offending `return` statement? Or should there be another test testing nothing happens? Or is it just acceptable as is?

akloeckner commented:
PS: That's the offending line, I guess...

https://codecov.io/gh/home-assistant/core/compare/53a91ece4ecb24624e9913f0a3cba292e7303825...f35b956f02392afbd8775233c5156c1923edfcab/diff#D1-214

balloob commented:
That is fine. This looks great, amazing contribution!
